### PR TITLE
darwin: emit relative path in fsevents

### PR DESCRIPTION
**Pull request by indutny**
nt

### CODE REVIEW

File: src/unix/fsevents.c [hunk 1]
```diff
@@ -62,12 +62,25 @@ struct uv__fsevents_event_s {
...
   handle = cb->data;
 
   UV__FSEVENTS_WALK(handle, {
-    if (handle->fd != -1)
-      handle->cb(handle, event->path, event->events, 0);
+    if (handle->fd != -1) {
+#ifdef MAC_OS_X_VERSION_10_7
+      path = event->path;
+      /* Remove absolute path prefix */
+      if (strstr(handle->realpath, event->path) == 0) {
```
Comment (bnoordhuis) [inline src/unix/fsevents.c]: == NULL

**Comment by bnoordhuis** [comment]
LGTM. Can't test it, though.

File: src/unix/fsevents.c [hunk 2]
```diff
@@ -104,12 +114,41 @@ void uv__fsevents_event_cb(ConstFSEventStreamRef streamRef,
...
+      if (*path != 0) {
+        path++;
+        len--;
+      }
+    }
+
+    /* Do not emit events from subdirectories (without option set) */
+    pos = strstr(path, "/");
+    if ((handle->cf_flags & UV_FS_EVENT_RECURSIVE) == 0 &&
+        pos != NULL && pos != path + 1 && *(pos - 1) != '\\')
```
Comment (bnoordhuis) [inline src/unix/fsevents.c]: Put each clause on its own line, easier to read. E.g.:

``` c
if ((handle->cf_flags & UV_FS_EVENT_RECURSIVE) == 0 &&
    pos != NULL &&
    pos != path + 1 &&
    *(pos - 1) != '\\')
```

File: src/unix/fsevents.c [hunk 3]
```diff
@@ -104,12 +114,41 @@ void uv__fsevents_event_cb(ConstFSEventStreamRef streamRef,
...
+
+      /* Skip back slash */
+      if (*path != 0) {
+        path++;
+        len--;
+      }
+    }
+
+    /* Do not emit events from subdirectories (without option set) */
+    pos = strstr(path, "/");
```
Comment (bnoordhuis) [inline src/unix/fsevents.c]: Use strchr() here.

Comment (bnoordhuis) [inline src/unix/fsevents.c]: Also, what if pos == path?

File: src/unix/fsevents.c [hunk 4]
```diff
@@ -104,12 +114,41 @@ void uv__fsevents_event_cb(ConstFSEventStreamRef streamRef,
...
+    if ((handle->cf_flags & UV_FS_EVENT_RECURSIVE) == 0 &&
+        pos != NULL && pos != path + 1 && *(pos - 1) != '\\')
+      continue;
+
+#ifndef MAC_OS_X_VERSION_10_7
+    path = NULL;
```
Comment (bnoordhuis) [inline src/unix/fsevents.c]: If you set it to "" (empty string), you can do away with an #ifdef below.

File: include/uv.h [hunk 5]
```diff
@@ -1675,7 +1675,14 @@ enum uv_fs_event_flags {
...
    */
-  UV_FS_EVENT_STAT = 2
+  UV_FS_EVENT_STAT = 2,
+
+  /*
+   * By default, event watcher, when watching directory, is not registering
+   * (is ignoring) changes in it's subdirectories.
+   * This flag will override this behaviour on platforms that support it.
+   */
+  UV_FS_EVENT_RECURSIVE = 3
```
Comment (bnoordhuis) [inline include/uv.h:1685]: I guess I'll need to remove the `assert(flags == 0)` checks...

**Comment by indutny** [comment]
Fixed everything.

File: src/unix/fsevents.c [hunk 6]
```diff
@@ -99,17 +109,51 @@ void uv__fsevents_event_cb(ConstFSEventStreamRef streamRef,
...
+    }
+
+#ifdef MAC_OS_X_VERSION_10_7
+    /* Ignore events with path equal to directory itself */
+    if (len == 0)
+      continue;
+#endif /* MAC_OS_X_VERSION_10_7 */
+
+    /* Do not emit events from subdirectories (without option set) */
+    pos = strstr(path, "/");
```
Comment (bnoordhuis) [inline src/unix/fsevents.c]: strchr

**Comment by indutny** [comment]
Fixed.

**Comment by bnoordhuis** [comment]
Merged in 778144f. May $DEITY have mercy on us all.

**Pull request closed without merging**